pkg/endpoint/restore.go

```diff
@@ -222,28 +223,72 @@ func (e *Endpoint) restoreIdentity() error {
 	l, _ := labelsfilter.Filter(e.OpLabels.AllLabels())
 	e.runlock()
 
-	allocateCtx, cancel := context.WithTimeout(context.Background(), option.Config.KVstoreConnectivityTimeout)
-	defer cancel()
-	identity, _, err := e.allocator.AllocateIdentity(allocateCtx, l, true)
+	// Getting the ep's identity while we are restoring should block the
+	// restoring of the endpoint until we get its security identity ID.
+	// If the endpoint is removed, this controller will cancel the allocator
+	// requests.
+	controllerName := fmt.Sprintf("restoring-ep-identity (%v)", e.ID)
+	var (
+		identity *identity.Identity
+		gotIdty  = make(chan struct{})
+	)
+	e.UpdateController(controllerName,
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) (err error) {
+				allocateCtx, cancel := context.WithTimeout(ctx, option.Config.KVstoreConnectivityTimeout)
+				defer cancel()
+				identity, _, err = e.allocator.AllocateIdentity(allocateCtx, l, true)
+				if err != nil {
+					return err
+				}
+				close(gotIdty)
+				return nil
+			},
+		})
 
-	if err != nil {
-		scopedLog.WithError(err).Warn("Unable to restore endpoint")
-		return err
+	// Wait until we either get an identity or the endpoint is removed or
+	// deleted from the node.
+	select {
+	case <-e.aliveCtx.Done():
+		return ErrNotAlive
+	case <-gotIdty:
 	}
 
 	// Wait for initial identities and ipcache from the
 	// kvstore before doing any policy calculation for
 	// endpoints that don't have a fixed identity or are
 	// not well known.
 	if !identity.IsFixed() && !identity.IsWellKnown() {
-		identityCtx, cancel := context.WithTimeout(context.Background(), option.Config.KVstoreConnectivityTimeout)
-		defer cancel()
+		// Getting the initial global identities while we are restoring should
+		// block the restoring of the endpoint.
+		// If the endpoint is removed, this controller will cancel the allocator
+		// WaitForInitialGlobalIdentities function.
```

Comment:
How is this step implemented? I see the exit case below to continue if `aliveCtx` is cancelled, but I don't see cleanup of the new controller.

Reply:
@joestringer all controllers in `e.controllers` are cleaned up when the endpoint is disconnected. Is part of the disconnect function.